### PR TITLE
Not so mini dev branch: various improvements 

### PR DESCRIPTION
**Pull request by vict0rsch**
Main additions:

* More comet logging: 
  ```yaml
  comet:
    display_size: 20
    rows_per_log: 5 # number of samples (rows) in a logged grid image. Number of total logged images: display_size // rows_per_log
  ```
* Better `s` image logging with `utils.get_display_indices`:
  ```python
	...
    if domain == "s":
        with temp_np_seed(123): # deterministic random
            display_indices = np.random.permutation(length)[:dsize]
	...
  ```

* **New `SegmentationDecoder`**

* New option: `data.max_samples=n` to limit the size of the dataset (for instance for dev purposes, `n=-1` for no limitation)

* Add `tqdm` logging

* Add `self.G.eval()` and `self.D.eval()` for the inference mode, especially for batch norm and dropout

* v0 for an almost automatic `resume.py`

* **hyper parameter search**

* Automatic comet experiment continuation:
  ```python
  ...
  with comet_previous_path.open("r") as f:
      url = f.read().strip()
      comet_previous_id = comet_id_from_url(url)
  ...
  exp = ExistingExperiment(previous_experiment=comet_previous_id, **comet_kwargs)
  ...
  ```

### CODE REVIEW

File: omnigan/trainer.py [hunk 1]
```diff
@@ -406,39 +426,45 @@ def log_comet_images(self, mode, domain):
...
-                        if update_task == "s":
-                            if domain == "s":
-                                target = (
-                                    decode_segmap_unity_labels(target, domain, True)
-                                    .float()
-                                    .to(self.device)
-                                )
-                            prediction = (
-                                decode_segmap_unity_labels(prediction, domain, False)
+                    if update_task == "x":
```
Comment (vict0rsch) [inline omnigan/trainer.py:429]: instead of having everything under `if task != "x"` I changed to `if update_task == "x": continue` to have less indentation (and more clarity)

**Review by sashavor — APPROVED** [review]

**Review by tianyu-z — APPROVED** [review]

File: omnigan/trainer.py [hunk 2]
```diff
@@ -536,10 +583,16 @@ def train(self):
         for self.logger.epoch in range(
             self.logger.epoch, self.logger.epoch + self.opts.train.epochs
         ):
-            # self.infer()
             self.run_epoch()
+            self.G.eval()
+            self.D.eval()
             self.infer(verbose=1)
-            if self.logger.epoch % self.opts.train.save_n_epochs == 0:
+            self.G.train()
```
Comment (alexrey88) [inline omnigan/trainer.py:590]: Is it necessary to call model.train() again here if you already do it at the beginning of run_epoch() ?

File: omnigan/trainer.py [hunk 3]
```diff
@@ -379,10 +393,14 @@ def run_epoch(self):
             self.log_step_time(step_time)
 
         for d in self.opts.domains:
```
Comment (alexrey88) [inline omnigan/trainer.py:395]: Why do you call model.eval() before every log and model.train() right after, instead of just calling model.eval() once before the for loop and calling model.train() once after the if ? I guess there's a reason, just wondering what it is :)

File: omnigan/trainer.py (hunk 2 above)
Comment (vict0rsch) [reply omnigan/trainer.py:590]: No you're right :) I think I added it inside run_epoch afterwards, by fear that if would be forgotten beforehand down the road if we re-write things. It would be a real pain in the ass to train in eval mode :p

**Comment by vict0rsch** [comment]
@alexrey88 I refactored the evaluation as you pointed out, do you see any other weird things?

**Comment by alexrey88** [comment]
> @alexrey88 I refactored the evaluation as you pointed out, do you see any other weird things?

All seems good :)

**Review by alexrey88 — APPROVED** [review]
